### PR TITLE
[fix] Workaround Classy doing nasty things with optimizers

### LINKED ISSUE #236: [ShardedGradScaler] Handle optimizers not being OSS instances
## 🚀 Feature
Some frameworks (Classy for instance) change the optimizers to enable different non-pytorch-compliant features. ShardedGradScaler asserts on that

## Motivation
Enable ShardedDDP + AMP on Classy-like frameworks

## Pitch
Remove the assert, replace by a one time warning

## Alternatives
Not doing anything

## Additional context
@mannatsingh

### PR DESCRIPTION
**Pull request by blefaudeux**
# Before submitting

- [x] Was this discussed/approved via a Github issue? (no need for typos, doc improvements)
- [x] Did you read the [contributor guideline](https://github.com/facebookresearch/fairscale/blob/master/CONTRIBUTING.md)?
- [ ] Did you make sure to update the docs?
- [ ] Did you write any new necessary tests?

## What does this PR do?
Fixes #236

## PR review
Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

**Comment by blefaudeux** [comment]
cc @prigoyal, needed for the ShardedDDP + AMP run that I shared late last week

File: fairscale/optim/grad_scaler.py [hunk 1]
```diff
@@ -31,9 +32,16 @@ class ShardedGradScaler(TorchGradScaler):
 
     def __init__(self) -> None:
         super().__init__()
+        self.warning = True
```
Comment (min-xu-ai) [inline fairscale/optim/grad_scaler.py:35]: change to "warned"?

**Review by min-xu-ai — APPROVED** [review]
fair enough!

File: fairscale/optim/grad_scaler.py (hunk 1 above)
Comment (blefaudeux) [reply fairscale/optim/grad_scaler.py]: it's "not warned" right now, renaming to "display_warning" if it's ok for you, to be super explicit ?

**Comment by prigoyal** [comment]
curious question: why do we need this fix?

**Comment by blefaudeux** [comment]
> curious question: why do we need this fix?

Classy wraps the optimizer in a dedicated class, which does not inherit from the actual optimizer you requested (see https://github.com/facebookresearch/ClassyVision/blob/0673482cb41683c72d0630b4765122cccef074b9/classy_vision/optim/classy_optimizer.py#L47, the attributes are passed on so that it pseudo-inherits but the type is different), so this assert would fail

**Pull request merged**